### PR TITLE
Add documentation of StaticFinalFieldFolding

### PR DESCRIPTION
**Pull request by liqunl**
Part of #11974

[ci skip]

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@0xdaryl May I have your review?

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 1]
```diff
@@ -30,6 +30,63 @@
 
 namespace TR { class NodeChecklist; }
 
+/** \brief
+ *
+ * This opt walks the trees and folds direct loads of static final field.
+ *
+ * \details
+ *
+ * This opt walks the trees and tries to fold direct loads of static final field.
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:39]: field -> fields

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 2]
```diff
@@ -30,6 +30,63 @@
...
+ * For primitive type, it will be folded into a primitive constant. For reference
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:40]: For a primitive type, ...

For a reference type, ...

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 3]
```diff
@@ -30,6 +30,63 @@
...
+ * type, a known object index is assigned to the referenced object, the symref on
+ * the load node is improved to one with the known object index.
+ *
+ * There are two types of static final fields this opt trying to fold:
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:44]: is trying to fold

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 4]
```diff
@@ -30,6 +30,63 @@
...
+ *     these fields are not expected to be modified once initialized.
+ *     `J9::TransformUtil::foldFinalFieldsIn` contains a list of such classes.
+ *
+ *  2. Other fields that may be modified through reflection or Unsafe APIs.
+ *
+ * For 1, they're always directly folded into a constant. The folding doesn't require any
+ * analysis or protection.
+ *
+ * For 2, they can only be folded with the assumption that they're not to be
+ * modified. However, when modification happens, there JIT code must not execute the code
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:55]: I think:  "...happens, the JIT code with the invalid folded value must not be executed."

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 5]
```diff
@@ -30,6 +30,63 @@
...
+ * with the invalid folded value. Thus, the folding of type 2 final statics requires guards.
+ * We use OSR guard to protect ourselves in such situation. When modification happens, the
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:57]: guard -> guards

situation -> situations

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp [hunk 6]
```diff
@@ -30,6 +30,63 @@
...
+ * OSR guard will be patched to induce OSR transition to the interpreter.
+ *
+ * To be able to do guarded static final field folding (to fold the second type of fields),
+ * we have to make sure that an OSR guard can be added after any yield point, place where
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:61]: place -> placed      ?

**Review by 0xdaryl — COMMENTED** [review]
Thanks for writing this.  Only a few minor typo corrections from me.

**Comment by liqunl** [comment]
@0xdaryl Made some changes according to your comments, could you review again?

**Comment by 0xdaryl** [comment]
Thanks.  I'll wait for @vijaysun-omr before merging.

**Comment by vijaysun-omr** [comment]
Can it be clearly stated whether both 1 and 2 foldings are enabled by default ? If so, at which opt levels or under which conditions ?

**Comment by vijaysun-omr** [comment]
I am fine with the rest of the document here (thanks @liqunl for the well written description) and once my prior comment is addressed please feel free to merge it without waiting for any formal approval from me.

**Comment by liqunl** [comment]
Added the conditions when the two types of folding happen to the end of the documentation. Also made a correction to the statement that the folding util can be called anytime in the optimizer. It can only be called before OSRGuardInsertion.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**